src/inmanta/server/services/userservice.py

```diff
@@ -0,0 +1,110 @@
+
+    Contact: code@inmanta.com
+"""
+import logging
+
+import nacl.exceptions
+import nacl.pwhash
+from inmanta import const, data, protocol
+from inmanta.data import model
+from inmanta.protocol import common, exceptions
+from inmanta.server import SLICE_TRANSPORT, SLICE_USER
+from inmanta.server import protocol as server_protocol
+
+LOGGER = logging.getLogger(__name__)
+
+
+class UserService(server_protocol.ServerSlice):
+    """Slice for managing users"""
+
+    def __init__(self) -> None:
+        super().__init__(SLICE_USER)
+
+    def get_depended_by(self) -> list[str]:
+        return [SLICE_TRANSPORT]
+
+    @protocol.handle(protocol.methods_v2.list_users)
+    async def list_users(self) -> list[model.User]:
+        return [user.to_dao() for user in await data.User.get_list()]
+
+    @protocol.handle(protocol.methods_v2.add_user)
+    async def add_user(self, username: str, password: str) -> model.User:
+        # check if the user already exists
+        existing = await data.User.get_one(username=username)
+        if existing:
+            raise exceptions.Conflict(f"A user with name {username} already exists.")
+
+        # hash the password
+        pw_hash = nacl.pwhash.str(password.encode())
+
+        # insert the user
+        user = data.User(
+            username=username,
+            password_hash=pw_hash.decode(),
+            enabled=True,
+            auth_method="password",
+        )
+        await user.insert()
+        return user.to_dao()
+
+    @protocol.handle(protocol.methods_v2.delete_user)
+    async def delete_user(self, username: str) -> None:
+        user = await data.User.get_one(username=username)
+        if user is None:
+            raise exceptions.NotFound(f"User with name {username} does not exist.")
+
+        await user.delete()
+
+    @protocol.handle(protocol.methods_v2.set_password)
+    async def set_password(self, username: str, password: str) -> None:
+        # check if the user already exists
+        user = await data.User.get_one(username=username)
+        if not user:
+            raise exceptions.NotFound(f"User with name {username} does not exist.")
+
+        # hash the password
+        pw_hash = nacl.pwhash.str(password.encode())
+
+        # insert the user
+        await user.update_fields(password_hash=pw_hash.decode())
+
+    @protocol.handle(protocol.methods_v2.login)
+    async def login(self, username: str, password: str) -> common.ReturnValue[model.LoginReturn]:
+        # check if the user exists
+        user = await data.User.get_one(username=username)
+        if not user:
+            raise exceptions.UnauthorizedException()
+
+        if not user.password_hash:
+            raise exceptions.UnauthorizedException()
+
+        try:
+            nacl.pwhash.verify(user.password_hash.encode(), password.encode())
+        except nacl.exceptions.InvalidkeyError:
+            raise exceptions.UnauthorizedException()
+
+        # TODO: set an expire
```

Comment:
TODO?

Reply:
I'll create tickets for all the todo's right before I merge this. I think this was the plan of this ticket?
`There are still many todo's of which many are actually a ticket on their own.`

Reply:
I would create the tickets and remove the TODO's from the PR. The goal of the review is also the verify that none of these TODO's are left behind. Now this is not possible and I don't see a reason to delay the creation of these tickets until right before the merge of the PR.

Reply:
https://app.zenhub.com/workspaces/core-5a7c152b04a1652024289b7b/issues/gh/inmanta/inmanta-core/5643